libnmstate/nm/checkpoint.py

```diff
@@ -124,6 +126,23 @@ def create(self):
                 raise NMCheckPointPermissionError(str(e))
             raise NMCheckPointCreationError(str(e))
 
+        GLib.timeout_add(
+            timeout * 0.5,
+            self._refresh_checkpoint_timeout,
+            (self._dbuspath, timeout),
+        )
+
+    def _refresh_checkpoint_timeout(self, data):
+        dbuspath, timeout = data
+        nm_client = client()
+        if nm_client:
```

Comment:
After investigating I've discovered that sometimes after finishing all the actions and quitting the mainloop, `nm_client` is `None`. I am not sure but I think this happens because the GLib timeout does not are about the NM mainloop.

So, right now, I am not doing nothing. If `nm_client` is `None` it means mainloop quit and therefore I am not refreshing the timeout (because I can't do it indeeed).

I am not 100% sure about this. Thanks!

Reply:
@cathay4t can you help? When is `nm_client` supposed to be `None`?

Reply:
It's a long story:

 * Activation finished, `_delete_client()` invoked, which then invoke `_nmclient.delete_client()`.
 * In `_nmclient.delete_client()` it set `self._client` as None.  iterating the main context to wait NM.Client clean up.
 * The main context got event from timeout, then invoke `_refresh_checkpoint_timeout()`. So that time, the `_nmclient` is not None while `_nmclient.client` is None.

That's why.

You code has many problems:
 * Never do `GLib.timeout_destroy()`.
 * Created a lot lot timeout event by recursively calling `GLib.timeout_add()`.
 * Need correct handling above problem.